File: src/main/java/com/university/services/impl/PrepaymentReportServiceImpl.java
package com.university.services.impl;

import com.university.entities.PrepaymentReport;
import org.springframework.beans.factory.annotation.Autowired;
import com.university.repositories.PrepaymentReportRepository;
import com.university.services.PrepaymentReportService;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
public class PrepaymentReportServiceImpl implements PrepaymentReportService {

    private final PrepaymentReportRepository prepaymentReportRepository;

    @Autowired
    public PrepaymentReportServiceImpl(PrepaymentReportRepository prepaymentReportRepository) {
        this.prepaymentReportRepository = prepaymentReportRepository;
    }

    @Override
    public void addPrepaymentReport(PrepaymentReport prepaymentReport) {
        prepaymentReportRepository.saveAndFlush(prepaymentReport);
    }

    @Override
    public void removePrepaymentReport(int id) {
        prepaymentReportRepository.delete(id);
    }

    @Override
    public PrepaymentReport getPrepaymentReportById(int id) {
        return prepaymentReportRepository.findOne(id);
    }

    @Override
    public List<PrepaymentReport> getPrepaymentReportsByDate(Optional<LocalDate> date) {
        if (date.isPresent()) {
            return prepaymentReportRepository.findPrepaymentReportsByPreparationDateBetween(LocalDate.of(date.get().getYear(), date.get().getMonth(), 1),
                    LocalDate.of(date.get().getYear(), date.get().getMonth(), date.get().lengthOfMonth()));
        } else {
            return prepaymentReportRepository.findAll();
        }
    }

    @Override
    public PrepaymentReport findPrepaymentReportByTravelAllowance_Id(int travelAllowanceId) {
        return prepaymentReportRepository.findPrepaymentReportByTravelAllowance_Id(travelAllowanceId);
    }

}

File: src/main/kotlin/com/university/services/impl/PrepaymentReportServiceImpl.kt
package com.university.services.impl

import com.university.entities.PrepaymentReport
import com.university.repositories.PrepaymentReportRepository
import com.university.services.PrepaymentReportService
import org.springframework.stereotype.Service
import java.time.LocalDate

@Service
class PrepaymentReportServiceImpl (private val prepaymentReportRepository: PrepaymentReportRepository) : PrepaymentReportService {

    override fun save(prepaymentReport: PrepaymentReport) {
        prepaymentReportRepository.saveAndFlush(prepaymentReport)
    }

    override fun removeById(id: Int) {
        prepaymentReportRepository.deleteById(id)
    }

    override fun findByDate(date: LocalDate?): List<PrepaymentReport> {
        return if (date != null) {
            prepaymentReportRepository.findPrepaymentReportsByPreparationDateBetween(LocalDate.of(date.year, date.month, 1),
                    LocalDate.of(date.year, date.month, date.lengthOfMonth()))
        } else {
            prepaymentReportRepository.findAll()
        }
    }
}
